# NLP: Tokenization,Sequences,Padding

Data: Sarcasm Detection using Hybrid Neural Network
Rishabh Misra, Prahal Arora
Arxiv, August 2019

Load the data

In [1]:
def parseJson(fname):
    for line in open(fname, 'r'):
        yield eval(line)

In [2]:
data = list(parseJson('Sarcasm_Headlines_Dataset.json'))

In [3]:
data

[{'is_sarcastic': 1,
  'headline': 'thirtysomething scientists unveil doomsday clock of hair loss',
  'article_link': 'https://www.theonion.com/thirtysomething-scientists-unveil-doomsday-clock-of-hai-1819586205'},
 {'is_sarcastic': 0,
  'headline': 'dem rep. totally nails why congress is falling short on gender, racial equality',
  'article_link': 'https://www.huffingtonpost.com/entry/donna-edwards-inequality_us_57455f7fe4b055bb1170b207'},
 {'is_sarcastic': 0,
  'headline': 'eat your veggies: 9 deliciously different recipes',
  'article_link': 'https://www.huffingtonpost.com/entry/eat-your-veggies-9-delici_b_8899742.html'},
 {'is_sarcastic': 1,
  'headline': 'inclement weather prevents liar from getting to work',
  'article_link': 'https://local.theonion.com/inclement-weather-prevents-liar-from-getting-to-work-1819576031'},
 {'is_sarcastic': 1,
  'headline': "mother comes pretty close to using word 'streaming' correctly",
  'article_link': 'https://www.theonion.com/mother-comes-pretty-

In [4]:
len(data)

28619

In [5]:
data[0]

{'is_sarcastic': 1,
 'headline': 'thirtysomething scientists unveil doomsday clock of hair loss',
 'article_link': 'https://www.theonion.com/thirtysomething-scientists-unveil-doomsday-clock-of-hai-1819586205'}

The dataset consists of 28,619 samples

Each sample in the dataset is a dictionary consisting of 3 key value pairs

'is_sarcastic': label

'headline': text

'article_link': url link to the article

In [6]:
labels=[]
headlines=[]
article_link=[]

In [7]:
for item in data:
    labels.append(item['is_sarcastic'])
    headlines.append(item['headline'])
    article_link.append(item['article_link'])

In [8]:
labels

[1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 1,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 1,
 0,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 1,
 0,
 1,
 0,
 1,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 1,


In [9]:
headlines

['thirtysomething scientists unveil doomsday clock of hair loss',
 'dem rep. totally nails why congress is falling short on gender, racial equality',
 'eat your veggies: 9 deliciously different recipes',
 'inclement weather prevents liar from getting to work',
 "mother comes pretty close to using word 'streaming' correctly",
 'my white inheritance',
 '5 ways to file your taxes with less stress',
 "richard branson's global-warming donation nearly as much as cost of failed balloon trips",
 'shadow government getting too large to meet in marriott conference room b',
 'lots of parents know this scenario',
 'this lesbian is considered a father in indiana (and an amazing one at that)',
 "amanda peet told her daughter sex is 'a special hug'",
 'what to know regarding current treatments for ebola',
 "chris christie suggests hillary clinton was to blame for boko haram's kidnapping of hundreds of schoolgirls",
 'ford develops new suv that runs purely on gasoline',
 'uber ceo travis kalanick step

In [10]:
import tensorflow as tf

In [11]:
from tensorflow import keras
from tensorflow.keras.preprocessing.text import Tokenizer

In [12]:
from tensorflow.keras.preprocessing.sequence import pad_sequences

Tokenization

In [13]:
tokenizer=Tokenizer(oov_token='OOV')

In [14]:
tokenizer.fit_on_texts(headlines)

In [15]:
word_index=tokenizer.word_index

In [16]:
word_index

{'OOV': 1,
 'to': 2,
 'of': 3,
 'the': 4,
 'in': 5,
 'for': 6,
 'a': 7,
 'on': 8,
 'and': 9,
 'with': 10,
 'is': 11,
 'new': 12,
 'trump': 13,
 'man': 14,
 'at': 15,
 'from': 16,
 'about': 17,
 'by': 18,
 'after': 19,
 'you': 20,
 'this': 21,
 'out': 22,
 'up': 23,
 'be': 24,
 'as': 25,
 'that': 26,
 'it': 27,
 'how': 28,
 'not': 29,
 'he': 30,
 'his': 31,
 'are': 32,
 'your': 33,
 'just': 34,
 'what': 35,
 'all': 36,
 'who': 37,
 'has': 38,
 'will': 39,
 'report': 40,
 'into': 41,
 'more': 42,
 'one': 43,
 'have': 44,
 'year': 45,
 'over': 46,
 'why': 47,
 'day': 48,
 'u': 49,
 'area': 50,
 'woman': 51,
 'can': 52,
 's': 53,
 'says': 54,
 'donald': 55,
 'time': 56,
 'first': 57,
 'like': 58,
 'no': 59,
 'her': 60,
 'get': 61,
 'off': 62,
 'old': 63,
 "trump's": 64,
 'life': 65,
 'now': 66,
 'people': 67,
 "'": 68,
 'an': 69,
 'house': 70,
 'still': 71,
 'obama': 72,
 'white': 73,
 'back': 74,
 'make': 75,
 'was': 76,
 'than': 77,
 'women': 78,
 'if': 79,
 'down': 80,
 'when': 81,
 'i'

In [17]:
len(word_index)

30885

Generate sequences

In [18]:
sequences=tokenizer.texts_to_sequences(headlines)

In [19]:
sequences

[[16004, 355, 3167, 7474, 2644, 3, 661, 1119],
 [7475, 1775, 758, 3168, 47, 239, 11, 1844, 1048, 8, 1528, 2154, 1845],
 [863, 33, 11427, 262, 16005, 647, 1484],
 [11428, 1594, 6418, 5665, 16, 143, 2, 152],
 [472, 495, 322, 922, 2, 572, 550, 16006, 6419],
 [83, 73, 11429],
 [84, 280, 2, 4182, 33, 2515, 10, 573, 1356],
 [1846, 16007, 777, 5063, 3169, 1120, 25, 227, 25, 1433, 3, 1434, 3359, 4183],
 [2961, 292, 143, 139, 1709, 2, 425, 5, 16008, 1595, 356, 1357],
 [2792, 3, 160, 115, 21, 7476],
 [21, 2645, 11, 3360, 7, 426, 5, 3580, 9, 69, 1006, 43, 15, 26],
 [7477, 11430, 648, 60, 536, 201, 11, 1358, 626, 16009],
 [35, 2, 115, 11431, 1710, 11432, 6, 1049],
 [574,
  1596,
  953,
  149,
  85,
  76,
  2,
  1776,
  6,
  9021,
  16010,
  7478,
  3,
  1149,
  3,
  11433],
 [1529, 4184, 12, 6420, 26, 1530, 11434, 8, 9022],
 [1318, 379, 7479, 9023, 5064, 80, 16, 13, 1711, 7480, 2155],
 [50, 433, 1150, 5666, 9, 2156, 2793, 3, 16011, 3361],
 [50, 14, 323, 100, 3, 31, 2261, 18, 11435],
 [537, 59, 411

Generated sequences are of different length, in order to feed the sequences into the neaural network, pad the sequences with zeros at the end of the sequences.

Now the length of the sequences is equal to the longest sequence in the list of sequences

In [20]:
padded=pad_sequences(sequences,padding='post')

In [21]:
padded.shape

(28619, 152)

In [22]:
padded

array([[16004,   355,  3167, ...,     0,     0,     0],
       [ 7475,  1775,   758, ...,     0,     0,     0],
       [  863,    33, 11427, ...,     0,     0,     0],
       ...,
       [    4,   100,   629, ...,     0,     0,     0],
       [ 1870,  1313,  3317, ...,     0,     0,     0],
       [  217,  3283,    21, ...,     0,     0,     0]])

Each headline text is now converted into a numeric vector of length 152.